### DISCUSSION TITLE
Allow a purchasable to define it's option attributes

### DISCUSSION THREAD

**Discussion by lukeholder** [craftcms/commerce#2429] [Ideas]
Purchasables should be able to define the following:

- What option attribute params they need when existing on a line item
- Validation information on those attributes (type (int, string), dropdown options etc)

For example the donation purchasable should be able to define the follow:

- donationAmount (number/currency) required

This would allow us to generate an add to cart UI specifically for that purchasable.

**Comment by AugustMiller** [comment]
What are the chances this leads to `LineItems` becoming Elements? There are a few DX things that would be improved as a result (like eager-loading), but perhaps most relevant would be the ability to piggy-back on the `FieldLayout` system…

In the interim, an explicit event for attaching extra `options` rules would be neat. To-date, I've just handled this by using the underlying `Model:: EVENT_DEFINE_RULES ` events:

```php
Event::on(
    LineItem::class,
    LineItem::EVENT_DEFINE_RULES,
    function (DefineRulesEvent $event) {
        $event->rules[] = [
            'options',
            function ($attribute, $params, $validator) use ($event) {
                // Check presence of certain attributes, etc!
            },
            'skipOnEmpty' => false,
            'when' => function ($li) {
                // Lazily determine whether the validator relevant:
                return $this->serviceName->computeSomeCriteria($li);
            }
        ];
    });
```

In thinking about this, though, I'm realizing I might misunderstand—is this just a way for built-in and third-party Purchasables to automatically register these, or a dynamic rule builder UI?

Maybe it's ultimately the same feature? (i.e. whatever UI is just a means for building extra `rules` that get attached to the `LineItem`?

**Comment by Anubarak** [comment]
Please don't transfer lineItems to elements. While it would have some benefits like eager loading (that's luckily solved by now) elements are much more resource heavy and consume an unnecessary amount of DB performance.
